freqtrade/exchange/binance_public_data.py

```diff
@@ -0,0 +1,210 @@
+"""
+Fetch daily-archived OHLCV data from https://data.binance.vision/
+"""
+
+import asyncio
+import datetime
+import io
+import logging
+import zipfile
+
+import aiohttp
+import pandas as pd
+from pandas import DataFrame
+
+from freqtrade.enums import CandleType
+from freqtrade.misc import chunks
+from freqtrade.util.datetime_helpers import dt_from_ts, dt_now
+
+
+logger = logging.getLogger(__name__)
+
+
+class BadHttpStatus(Exception):
+    """Not 200/404"""
+
+    pass
+
+
+async def fetch_ohlcv(
+    candle_type: CandleType,
+    pair: str,
+    timeframe: str,
+    since_ms: int,
+    until_ms: int | None,
+    stop_on_404: bool = False,
+) -> DataFrame:
+    """
+    Fetch OHLCV data from https://data.binance.vision/
+    :candle_type: Currently only spot and futures are supported
+    :param until_ms: `None` indicates the timestamp of the latest available data
+    :param stop_on_404: Stop to download the following data when a 404 returned
+    :return: the date range is between [since_ms, until_ms),
+        return None if no data available in the time range
+    """
+    if candle_type == CandleType.SPOT:
+        asset_type = "spot"
+    elif candle_type == CandleType.FUTURES:
+        asset_type = "futures/um"
+    else:
+        raise ValueError(f"Unsupported CandleType: {candle_type}")
+    symbol = symbol_ccxt_to_binance(pair)
+    start = dt_from_ts(since_ms)
```

Comment:
I think starting date handling is a key part - and it's currently unclear to me how you're handling the following (not necessarily tied to this line - but most likely close to it)

Assume i have existing data for a pair - which ends on 2024-10-25 at 11:35 am.
What will be the first file to be downloaded? 
How do you ensure the data ain't stored twice or contains holes (like - the rest of the 25th is missing completely)?
The same actually applies at the end of the timerange (where we assume "data stops here - let's use api from here onwards).

Reply:
>What will be the first file to be downloaded?

2024-10-25
https://github.com/freqtrade/freqtrade/blob/3c76af9dab59973e41f6cbde31356c3dfe4f56b5/freqtrade/exchange/binance_public_data.py#L125-L129
>How do you ensure the data ain't stored twice

I don't. The logic should be implemented by the caller of `get_historic_ohlcv`, so I just assumed If I return the  data of  timerange [since, until) (a left-closed right open timerange), it will work fine (I know the assumption might be wrong, ... and it is)

https://github.com/freqtrade/freqtrade/blob/3c76af9dab59973e41f6cbde31356c3dfe4f56b5/freqtrade/exchange/binance_public_data.py#L65-L68

>or contains holes (like - the rest of the 25th is missing completely)?

`fetch_ohlcv` has two modes with different assumptions:
1. `stop_on_404` is `False`: This mode we trust https://data.binance.vision provide the same data as rest API, i.e. a 404 error means there is no data in that day, so no holes.
2. `stop_on_404` is `True`: This mode we don't trust https://data.binance.vision, any 404 error means we should stop the following data download, and should fall back to rest API, so no holes.

https://github.com/freqtrade/freqtrade/blob/3c76af9dab59973e41f6cbde31356c3dfe4f56b5/freqtrade/exchange/binance_public_data.py#L111-L121

>The same actually applies at the end of the timerange (where we assume "data stops here - let's use api from here onwards)

https://github.com/freqtrade/freqtrade/blob/3c76af9dab59973e41f6cbde31356c3dfe4f56b5/freqtrade/exchange/binance.py#L148-L163

rest means REST API or rest ...

----------
I think this kind of corner cases should be tested by some real command line rather than plenty of mocks, if you think this PR worth to proceed on, I'd like to add some test cases.